File: pherix/compensators/identity.py
"""Identity & comms compensators.

  invite      → revoke_invite     (send an org/team invite → rescind it)
  grant_role  → revoke_role       (grant a permission → take it back)
  send_email  → (GATE, no comp)   (you cannot unsend an email — honest gate)

The first two are clean left-inverses keyed by a shared id. The third is
the catalog's worked example of *honesty about what cannot be undone*: an
email, once delivered, is in the recipient's inbox forever. There is no
semantic inverse, so the action is registered with **no compensator** and
``commit()`` gates on it — a human (or higher-trust policy) must call
``approve_irreversible()`` to let it through. The "undo" is the gate
itself: the chance to refuse *before* the irreversible thing happens.
"""

from __future__ import annotations

from pherix.core.tools import tool


def register_invite_revoke(client, *, resource: str = "identity"):
    """Register ``invite`` and its left-inverse ``revoke_invite``.

    ``client`` must expose::

        client.invite(invite_id, email, org) -> object
        client.revoke_invite(invite_id) -> object

    Reverses by ``invite_id`` — the caller mints it, so the invite and its
    revocation share the key the compensator receives in the args.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def revoke_invite(invite_id, email, org):
        return client.revoke_invite(invite_id)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="revoke_invite",
    )
    def invite(invite_id, email, org):
        return client.invite(invite_id, email, org)

    return invite, revoke_invite


def register_grant_revoke_role(client, *, resource: str = "identity"):
    """Register ``grant_role`` and its left-inverse ``revoke_role``.

    ``client`` must expose::

        client.grant_role(principal, role) -> object
        client.revoke_role(principal, role) -> object

    Reverses by ``(principal, role)`` — both carried in the args, so the
    compensator removes exactly the grant the action added.
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def revoke_role(principal, role):
        return client.revoke_role(principal, role)

    @tool(
        resource=resource,
        reversible=False,
        injects_handle=False,
        compensator="revoke_role",
    )
    def grant_role(principal, role):
        return client.grant_role(principal, role)

    return grant_role, revoke_role


def register_send_email_gate(client, *, resource: str = "identity"):
    """Register ``send_email`` with **no compensator** — it gates at commit.

    ``client`` must expose ``client.send_email(to, subject, body) -> object``.

    There is no honest left-inverse of a delivered email, so Pherix does not
    pretend one exists. The action stages, and ``commit()`` blocks until a
    human calls ``approve_irreversible(effect_id)``. Returns the action only
    (there is no compensator to return).
    """

    @tool(resource=resource, reversible=False, injects_handle=False)
    def send_email(to, subject, body):
        return client.send_email(to, subject, body)

    return send_email

File: pherix-ts/src/compensators/identity.ts
/**
 * Identity & comms compensators. Mirror of pherix/compensators/identity.py.
 *
 *   invite      → revokeInvite      (send an org/team invite → rescind it)
 *   grantRole   → revokeRole        (grant a permission → take it back)
 *   sendEmail   → (GATE, no comp)   (you cannot unsend an email — honest gate)
 *
 * The first two are clean left-inverses keyed by a shared id. The third is the
 * catalog's worked example of honesty about what cannot be undone: an email,
 * once delivered, is in the recipient's inbox forever. There is no semantic
 * inverse, so the action is registered with no compensator and `commit()` gates
 * on it — a human must call `approveIrreversible()`. The "undo" is the gate
 * itself: the chance to refuse *before* the irreversible thing happens.
 */

import { tool, type ToolWrapper } from "../tools.js";

export interface IdentityClient {
  invite(inviteId: string, email: string, org: string): unknown;
  revokeInvite(inviteId: string): unknown;
  grantRole(principal: string, role: string): unknown;
  revokeRole(principal: string, role: string): unknown;
  sendEmail(to: string, subject: string, body: string): unknown;
}

export interface InviteArgs extends Record<string, unknown> {
  inviteId: string;
  email: string;
  org: string;
}

/** Register `invite` and its left-inverse `revokeInvite`. Reverses by
 *  `inviteId` — the caller mints it, so action and revocation share the key. */
export function registerInviteRevoke(
  client: IdentityClient,
  resource = "identity",
): { invite: ToolWrapper<InviteArgs, unknown>; revokeInvite: ToolWrapper<InviteArgs, unknown> } {
  const revokeInvite = tool<InviteArgs>(
    resource,
    (args: InviteArgs) => client.revokeInvite(args.inviteId),
    { name: "revokeInvite", reversible: false, injectsHandle: false },
  );

  const invite = tool<InviteArgs>(
    resource,
    (args: InviteArgs) => client.invite(args.inviteId, args.email, args.org),
    { name: "invite", reversible: false, injectsHandle: false, compensator: "revokeInvite" },
  );

  return { invite, revokeInvite };
}

export interface RoleArgs extends Record<string, unknown> {
  principal: string;
  role: string;
}

/** Register `grantRole` and its left-inverse `revokeRole`. Reverses by
 *  `(principal, role)` — both carried in the args. */
export function registerGrantRevokeRole(
  client: IdentityClient,
  resource = "identity",
): { grantRole: ToolWrapper<RoleArgs, unknown>; revokeRole: ToolWrapper<RoleArgs, unknown> } {
  const revokeRole = tool<RoleArgs>(
    resource,
    (args: RoleArgs) => client.revokeRole(args.principal, args.role),
    { name: "revokeRole", reversible: false, injectsHandle: false },
  );

  const grantRole = tool<RoleArgs>(
    resource,
    (args: RoleArgs) => client.grantRole(args.principal, args.role),
    { name: "grantRole", reversible: false, injectsHandle: false, compensator: "revokeRole" },
  );

  return { grantRole, revokeRole };
}

export interface EmailArgs extends Record<string, unknown> {
  to: string;
  subject: string;
  body: string;
}

/** Register `sendEmail` with no compensator — it gates at commit. There is no
 *  honest left-inverse of a delivered email, so Pherix does not pretend one
 *  exists: the action stages and `commit()` blocks until a human calls
 *  `approveIrreversible(effectId)`. Returns the action only. */
export function registerSendEmailGate(
  client: IdentityClient,
  resource = "identity",
): ToolWrapper<EmailArgs, unknown> {
  return tool<EmailArgs>(
    resource,
    (args: EmailArgs) => client.sendEmail(args.to, args.subject, args.body),
    { name: "sendEmail", reversible: false, injectsHandle: false },
  );
}
